### DISCUSSION TITLE
[community poll] who is using pypdf?

### DISCUSSION THREAD

**Discussion by MartinThoma** [py-pdf/pypdf#683] [General]
Originally the following was to get an understanding of how PyPDF2 was used to ensure I don't break critical functionality. Now it's a continuous community feedback for pypdf :-)

Please share:

1) operating system
2) python version
3) what you use pypdf for 

Common examples of (3) are:

* Extracting Text (which language(s)? Which types of content, e.g. are there math formulas or tables?)
* Adding watermarks / stamps / overlays
* Combining PDFs
* Adding annotations
* Filling PDF forms

The more details, the better :-)

**Comment by ghost** [comment]
1. OS is RHEL 7 and 8

2. Python 2.7.5 (blame RHEL7) and 3.6+

3. I am using PyPDF2 (with a bunch of my own hacks) for populating PDF forms for both internal and client-facing purposes. Being able to set things like form field read/write, text colour (some of the fields are colour-coded based on a priority ranking), etc. are key things we use. I think I had to extend PyPDF2 to make this work, so I may be able to add these as a PR in future.

**Reply by MartinThoma** [reply]
I'm curious: Are you still using RHEL 7? Is Python 3.7 a choice now?

**Comment by hvbtup** [comment]
I'm using it on Windows and Linux, with Python 2.7 and with Python 3.6 and 3.10.
I made some little changes in my fork in order to support to an additional "background" argument in the various mergePage methods. The effect is that in mergePage, the new page can be used as a watermark ("under the existing content").

I noticed a bug in my fork just a few minutes ago. edit: Fixed

Feel free to include my changes here under the existing license if you like (but please wait until I've fixed that bug, which will happen next week).

I'm using PyPDF for two purposes:
a) concatenating PDFs
b) merging pages
We are also using commercial software for these purposes, but for "smaller" installations PyPDF2 suffices.

**Reply by hvbtup** [reply]
We are upgrading to Python 3.10/3.11, we'll no longer use Python 2.7 and 3.6.

**Comment by pubpub-zz** [comment]
I've used pypdf4 with some fixes, evolutions.
I used it also to extract/merge comments

**Comment by vteran93** [comment]
Hi, 

AWS Lambda with Serverless framework
python 3.8.10 using poetry
reading/writing/manipulating PDFs

**Comment by MartinThoma** [comment]


**Reply by clach04** [reply]
Dropping is more than reasonable, that being said many SBCs are on python 3.5.3 and compiling a later version is onerous. My recommendation is to branch or tag before dropping support and then document that in the readme for older python versions to make discovery easier.

**Comment by LightningMan711** [comment]
Windows whatever the current is (10 I think)
Python 3.8.2 (mostly; I still have code that needs 2.7)
Reading pages, manipulating bookmarks.

**Comment by LightningMan711** [comment]
I also must say that I am glad this project is being revived and updated.

**Comment by ghost** [comment]
> > Python 2.7.5 (blame RHEL7) and 3.6
> 
> @Joshua-IRT Uh, I was hoping that 3.6 and older isn't used any longer 🙈 Is [pyenv](https://github.com/pyenv/pyenv) an option for you?

Sorry, Python 3.6.8 is the main version on RHEL8. I have not tried pyenv on our systems yet, but can certainly look into it.

Since Python 3.6 is end-of-life, though, I can certainly understand the desire not to support it (and agree with it). I guess implement what is easy to do in 3.6, make a final release that supports it, then deprecate. I can continue to use my hacked-together version until I get around to setting up pyenv, at least.

**Comment by TZanke** [comment]
openSUSE Leap 15.x
python 3.6.x

- reading
- writing
- merging
- edit&create metadata with PyPDF2.generic.X

Nice to see progress again, we started using `PyMuPDF` for new code (if possible) cause PyPDF2's lack of fixed bugs.

**Comment by johns1c** [comment]
Windows 11
Python 3.8

1.  extracting text from received documents
2. extracting jpeg images and then running OCR to obtain text from scanned documents (HP  all in one)
3. viewing pdfs using a (much modified version )  of  Wx.lib.pdfviewer   

I am happy to contribute to the project if you  want help - for example

1. I have a collection of PDFs which cover a variety of  object types , fonts, encoding  schemes etc and  am willing  to carry out testing

2. I  have made some amendments to  my my own  fork of PyPDF2 which  I would like  to  be  reviewed and perhaps rolled back  into the  master, for example  one which  allows  text extraction using the  to_unicode objects (see Johns1c/PyPDF2) 

Chris

**Comment by nicksofn** [comment]
I use PyPDF3 but I'll say it is almost the same as PyPDF2 and if the community is moving back to PyPDF2 then it seems logical to migrate there too.

**Operating system:**
•	IBM i 7.3
•	Windows Server 2012 R2

**Python version:**
•	IBM i - 3.9.11-1
•	Windows Server 2012 R2 – 3.10.4

**What you use PyPDF2 for?**

Projects now and possibly in the future read PDF contents, split and merge.
Daily: take several PDFs and merge them for printing
Other: Split large PDFs and read contents of the resulting PDFs for proper categorization.

**Comment by zeshuaro** [comment]
1. **Operating system** - Python in Docker which I think it uses Debian, local dev on macOS
2. **Python** - 3.9.7
3. **What you use PyPDF2 for?** - Project is open source which you can check it out [here](https://github.com/zeshuaro/telegram-pdf-bot) if you want to, but uses PyPDF2 for:
    - Reading and Writing PDFs
    - Encrypt, decrypt, rotate and scale PDFs
    - Merge and split PDFs
    - Add watermark to PDFs
    - ~~Extract images from PDFs~~

**Reply by MartinThoma** [reply]
I've just added a few missing ones: Now all you mentioned is documented at https://pypdf2.readthedocs.io/ :-) 

https://github.com/zeshuaro/telegram-pdf-bot - interesting! I was thinking about an "awesome-python-pdf" List / Repo :thinking: 
How do people actually use that bot? Maybe I'm the wrong target group, but an image that shows the usage would be helpful to me :sweat_smile:

**Reply by zeshuaro** [reply]
Thanks for updating the documentation, that was very quick! 

I also realised that for extracting images from PDFs, I moved away from using PyPDF2 to using a CLI instead. I think I encountered some issues with PyPDF2 at some point, which made me look for an alternative. But if you're interested in the approach with PyPDF2, [here](https://github.com/zeshuaro/telegram-pdf-bot/pull/114/files#diff-5c2c3c29a12f86b2034caa79f7f0da35487d05ed5e65fae9d0b885f1e03b16f0) is the PR where I made the changes (I didn't come up with the approach using PyPDF2, think I found it somewhere on StackOverflow).

> I was thinking about an "awesome-python-pdf" List / Repo

Obviously I wouldn't mind if I can submit my project onto the list 😛  But I'm not sure how many Python PDF projects are there to be included onto the list.

> How do people actually use that bot?

People can use the bot through Telegram. Below is the stats captured in November last year. As you can probably guess, I didn't start capturing the stats from the beginning, so the "Estimated total tasks" is an approximation of the bot's usage since it was released.

Hopefully this will also give you some additional motivation on maintaining PyPDF2, as it's actually been used by hundreds of thousands of users indirectly 🙂 

![IMG_4347](https://user-images.githubusercontent.com/12210067/162609319-cbc9248c-2855-442a-9640-526e6029dcb5.PNG)

**Comment by htgoebel** [comment]
You may want to add pdfposter to the list.
https://pdfposter.readthedocs.io/en/stable/
https://pypi.org/project/pdftools-pdfposter/

**Comment by clach04** [comment]
I'm using it for a range of thing, across different Python versions. Main uses cases are:

* meta data extraction
* page extraction/split - then writing those pages to new PDFs

I use this for work and personal use cases.

I have one usage one Red Hat Enterprise Linux Server release 6.4 (Santiago) with Python 2.6.6 - BUT I do not need or expect maintenance for this to continue ;).

I have a bunch of usage with Python 2.7 and a range of 3.x. I have a bunch of SBCs (https://en.wikipedia.org/wiki/Single-board_computer) and Python 3.5 is the latest Python 3.x version so I'm still avoiding f-string (https://peps.python.org/pep-0498/) usage in my code for this reason. I have built later versions of Python on those platforms BUT it can take a few hours as they are low powered ARM boards.

Happy to answer questions if you have any :-)

Thanks again for working on this!

**Reply by MartinThoma** [reply]
Thanks for the feedback!

> Python 3.5 is the latest Python 3.x version

I guess you mean the latest for which a pre-built Python is available? It's astonishing to me that no newer version of Python is distributed for those SBCs. Are they that exotic?

To be honest, I have very little insight in non-standard architectures (At work we recently had the "fun" of M1 not having wheels for some packages we needed)

**Reply by clach04** [reply]
Yes, exactly I should been explicit. Pre-built bins of Python.

It looks like newer OS versions might have caught up, I think https://www.raspberrypi.com/news/raspberry-pi-os-debian-bullseye/ might be including 3.8 now.

Yeah, I hear you on the wheel pain - It is worth having a full dev chain so that pip can really install everything. Just need "disk" space and time :-)

**Comment by erenirmak** [comment]
I am working in the HR department and I am using PyPDF2 to process CV PDFs. I use it to extract education, job history, communication information extraction. Based on the extraction, I categorize CVs and recommend who should be interviewed with.

Sometimes, I receive CVs with images and have problems to process them. Sometimes I receive 200+ CVs daily, and it becomes hard to detect which one is not processed. I found a work around myself but still too much work needs to be done..

Currently, I am using Python 3.9.7.
Operating system: Windows 10 Pro 22H2

**Reply by MartinThoma** [reply]
Wow, that's a lot of CVs - and a very interesting application of PyPDF!

**Comment by mindphil** [comment]
1. Operating system: Windows 11
2. Python version: 3.13.5, 3.12
3. What you use pypdf for: reading PDF's and extracting certain strings (i.e. dates, digits, ect)

**Comment by johns1c** [comment]
Enviroment 

Mostly 

- Windows 11 
 - Python 3.8, 3.11, 3.12, 3.13 depending on application and generally with WxPython 
 - 
Also 
  Synology 7.1 (linux)
  Python 3.8 / 3.11 with Flask

Use 

Metadata extraction
1)    for scanned documents pypdf is used to extract the scanned images from the pdf wrapper prior to OCR and text processing
2)    for other documents the full text is  extracted and then processed to match certain data within it,  such as, company names, dates and subjcts.

I also use it to extract data from certain documents such as contract notes and invoices

I have used it in the past to display documents, rearrange pages, add and remove pictures but now use bespoke software to do thus 
 




(

**Comment by hoehermann** [comment]
I started this project at my place of work: https://github.com/hoehermann/pypdf_strreplace

It is ought to be a "search and replace" utility for changing text contents.

Context this project is being used in: At one point, I ended up with hundreds of documents which needed adjustments. A reporting tool had first been misconfigured, used for a short time, then decommissioned, leaving me with only the PDFs but no way of producing them anew.

* OS: Windows 11, Linux (Ubuntu, Arch)
* Python: 3.12 (whatever comes with the current LTS Ubuntu, newer on Arch)
